pkg/natdiscovery/response_handle.go

```diff
@@ -0,0 +1,88 @@
+/*
+© 2021 Red Hat, Inc. and others
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package natdiscovery
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+
+	"github.com/submariner-io/submariner/pkg/natdiscovery/proto"
+)
+
+func (nd *natDiscovery) handleResponseFromAddress(req *proto.SubmarinerNatDiscoveryResponse, addr *net.UDPAddr) error {
+	if req.GetSender() == nil || req.GetReceiver() == nil || req.GetReceivedSrc() == nil {
+		return errors.Errorf("received malformed response %#v", req)
+	}
+
+	if req.Response != proto.ResponseType_OK && req.Response != proto.ResponseType_SRC_MODIFIED {
+		var ok bool
+		var name string
+
+		if name, ok = proto.ResponseType_name[int32(req.Response)]; !ok {
+			name = fmt.Sprintf("%d", req.Response)
+		}
+
+		return errors.Errorf("remote endpoint %q responded with %q : %#v", req.Sender.EndpointId, name, req)
+	}
+
+	nd.Lock()
+	remoteNat, ok := nd.remoteEndpoints[req.GetSender().EndpointId]
+	nd.Unlock()
```

Comment:
We can't unlock here b/c we also need to protect access to the `remoteNat` instance below, especially since it mutates state. We might as well defer.

Reply:
oh right.